packages/util-dns/src/util/LinkedList.ts

```diff
@@ -0,0 +1,109 @@
+/**
+ * LinkedList implementation that implements {@link DnsCacheEntryCollection}
+ * for {@link HostEntry}.
+ * Can use any {@link NodeT} that conforms to the {@link LinkedListNode} interface.
+ * @internal
+ */
+export class LinkedList<NodeT extends LinkedListNode<NodeT>> {
```

Comment:
JS array is sufficiently optimizable serve as LL. Should be benchmarked if concerning, but I don't think we should have a custom LL implementation for a specific interface.

```ts
class DnsCacheEntryImpl<T> {
    private data: T[] = [];

    public get length(): number {
        return this.data.length;
    }

    public cycle(collection?: DnsCacheEntryImpl<T>) {
        const poll = this.data.shift();
        if (!poll) {
            throw new Error('collection is empty');
        }
        (collection ?? this).data.push(poll);
        return poll;
    }

    public append(item: T) {
        this.data.push(item);
    }

    public remove(item: T): T {
        const index = this.data.indexOf(item);
        const removed = this.data[index];
        this.data.splice(index, 1);
        return removed;
    }
}
```

Reply:
Resolved in https://github.com/aws/aws-sdk-js-v3/pull/4433/commits/0ac10ffe7f385e67c9693bb44f3391d4cd475eb4

Reply:
Also if anyone is curious, did a short analysis of LinkedList vs Array implementations of a `DnsCacheEntryCollection`: https://gist.github.com/syall/4495ee0b8f8648cfdb0e6430d3df45e1

At <100 entries (which is more aligned to reality):

- Arrays perform better for `append()` 
- Arrays perform slightly better for `cycle()`
- Arrays and LinkedLists have negligible differences in performance for `remove()`.

However, it should be noted that at scale (max of ~100,000 entries):

- Arrays perform better for `append()`
- LinkedLists perform better for `cycle()`
- LinkedLists perform better for `remove()`